File: TCP Split File/SplitFileServer_rust/src/SplitFileServer.rs
/* 20195914 */
/* Jeong eui chan */

use std::env;
use std::fs::File;
use std::io::{Read, Write, BufRead, BufReader};
use std::net::{TcpListener, TcpStream};
use std::thread;

// Function to handle each client connection
fn handle_client(mut stream: TcpStream) {
    let mut reader = BufReader::new(&stream);
    let mut command = String::new();
    
    // Read the command from the client
    if reader.read_line(&mut command).is_err() {
        eprintln!("Failed to read command from client");
        return;
    }

    // Split the command into parts
    let parts: Vec<&str> = command.trim().split_whitespace().collect();
    if parts.len() < 2 {
        eprintln!("Invalid command: {}", command);
        return;
    }

    let command_type = parts[0];
    let filename = parts[1];

    // Execute the appropriate action based on the command type
    match command_type {
        "put" => receive_file(&mut reader, filename),
        "get" => send_file(&mut stream, filename),
        _ => eprintln!("Unknown command: {}", command_type),
    }
}

// Function to receive a file from the client and save it
fn receive_file(reader: &mut BufReader<&TcpStream>, filename: &str) {
    let mut file = File::create(filename).expect("Failed to create file");
    let mut buffer = Vec::new();
    reader.read_to_end(&mut buffer).expect("Failed to read data");
    file.write_all(&buffer).expect("Failed to write data to file");
    println!("File received and saved as {}", filename);
}

// Function to read a file and send it to the client
fn send_file(stream: &mut TcpStream, filename: &str) {
    let mut file = match File::open(filename) {
        Ok(file) => file,
        Err(_) => {
            writeln!(stream, "Error: file {} does not exist", filename).unwrap();
            return;
        }
    };
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");
    stream.write_all(&buffer).expect("Failed to send data");
    println!("File sent: {}", filename);
}

// Main function to start the server and listen for incoming connections
fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: cargo run <port>");
        return;
    }

    let port = &args[1];
    let listener = TcpListener::bind(format!("0.0.0.0:{}", port)).expect("Failed to bind to port");
    println!("Server is listening on port {}", port);

    // Accept incoming connections and handle them in separate threads
    for stream in listener.incoming() {
        match stream {
            Ok(stream) => {
                thread::spawn(|| handle_client(stream));
            }
            Err(e) => eprintln!("Failed to accept connection: {}", e),
        }
    }
}

File: TCP_Split_File/SplitFileServer.go
/* 20195914 */
/* Jeong eui chan */

package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"strings"
)

// The main function is the entry point of the program.
func main() {
	// Check if the number of command-line arguments is 2. If not, print the usage message.
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run SplitFileServer.go <port>")
		return
	}

	// Get the port number and start a TCP listener.
	port := os.Args[1]
	listener, err := net.Listen("tcp", ":"+port)
	if err != nil {
		fmt.Println("Error starting server:", err)
		return
	}
	defer listener.Close()
	fmt.Println("Server is listening on port", port)

	// Infinite loop to accept connections.
	for {
		conn, err := listener.Accept()
		if err != nil {
			fmt.Println("Error accepting connection:", err)
			continue
		}
		// Handle each connection in a separate goroutine.
		go handleConnection(conn)
	}
}

// handleConnection handles each client connection.
func handleConnection(conn net.Conn) {
	defer conn.Close()
	reader := bufio.NewReader(conn)
	command, err := reader.ReadString('\n')
	if err != nil {
		fmt.Println("Error reading from connection:", err)
		return
	}

	// Trim and split the command into parts.
	command = strings.TrimSpace(command)
	parts := strings.Split(command, " ")
	if len(parts) < 2 {
		fmt.Println("Invalid command:", command)
		return
	}
	commandType := parts[0]
	filename := parts[1]

	// Execute the appropriate function based on the command type.
	switch commandType {
	case "put":
		receiveFile(reader, filename)
	case "get":
		sendFile(conn, filename)
	default:
		fmt.Println("Unknown command:", commandType)
	}
}

// receiveFile receives a file from the client and saves it.
func receiveFile(reader *bufio.Reader, filename string) {
	data, err := ioutil.ReadAll(reader)
	if err != nil {
		fmt.Println("Error reading file data:", err)
		return
	}
	err = ioutil.WriteFile(filename, data, 0644)
	if err != nil {
		fmt.Println("Error writing file:", err)
		return
	}
	fmt.Println("File received and saved as", filename)
}

// sendFile reads a file and sends it to the client.
func sendFile(conn net.Conn, filename string) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		if os.IsNotExist(err) {
			fmt.Fprintf(conn, "Error: file %s does not exist\n", filename)
		} else {
			fmt.Println("Error reading file:", err)
		}
		return
	}
	_, err = conn.Write(data)
	if err != nil {
		fmt.Println("Error sending file:", err)
	}
	fmt.Println("File sent:", filename)
}
